src/include/libpmemobj++/v.hpp

```diff
@@ -0,0 +1,174 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * Volatile resides on pmem property template.
+ */
+
+#ifndef PMEMOBJ_V_HPP
+#define PMEMOBJ_V_HPP
+
+#include <memory>
+
+#include "libpmemobj++/detail/common.hpp"
+#include "libpmemobj++/detail/volatile.hpp"
+
+namespace pmem
+{
+
+namespace obj
+{
+
+/**
+ * Volatile resides on pmem class.
+ *
+ * v class is a property-like template class that has to be used for all
+ * volatile variables that reside on persistent memory.
+ * This class ensures that the enclosed type is always properly initialized by
+ * always calling the class default constructor exactly once per instance of the
+ * application.
+ * This class has 8 bytes of storage overhead.
+ * @snippet doc_snippets/persistent.cpp v_property_example
+ */
+template <typename T>
+class v {
+
+public:
+	/**
+	 * Value constructor.
+	 *
+	 * Directly assigns a value to the underlying storage.
+	 *
+	 * @param _val const reference to the value to be assigned.
+	 */
+	v(const T &_val) noexcept : vlt{0}, val{_val}
+	{
+	}
+
+	/**
+	 * Defaulted constructor.
+	 */
+	v() = default;
+
+	/**
+	 * Assignment operator.
+	 */
+	v &
+	operator=(const v &rhs)
+	{
+		this_type(rhs).swap(*this);
+
+		return *this;
+	}
+
+	/**
+	 * Converting assignment operator from a different v<>.
+	 *
+	 * Available only for convertible types.
+	 *
+	 * @throw pmem::transaction_error when adding the object to the
+	 *	transaction failed.
+	 */
+	template <typename Y, typename = typename std::enable_if<
+				      std::is_convertible<Y, T>::value>::type>
+	v &
+	operator=(const v<Y> &rhs)
+	{
+		this_type(rhs).swap(*this);
+
+		return *this;
+	}
+
+	/**
+	 * Retrieves reference of the object.
+	 *
+	 * @return a reference to the object.
+	 *
+	 */
+	T &
+	get() noexcept
+	{
+		PMEMobjpool *pop = pmemobj_pool_by_ptr(this);
+		if (pop == NULL)
+			return this->val;
+
+		T *value = static_cast<T *>(pmemobj_direct_volatile(
+			pop, &this->vlt,
+			pmem::detail::instantiate_volatile_object<T>, NULL));
+
+		return *value;
+	}
+
+	/**
+	 * Conversion operator back to the underlying type.
+	 */
+	operator T() const noexcept
+	{
+		return this->get();
+	}
+
+	/**
+	 * Swaps two v objects of the same type.
+	 *
+	 * @throw pmem::transaction_error when adding the object to the
+	 *	transaction failed.
+	 */
+	void
+	swap(v &other)
+	{
+		std::swap(this->val, other.val);
+	}
+
+private:
```

Comment:
protected?

Reply:
This class is not meant for inheriting.

Reply:
Actually, val should be public or there should be simple getter for val.
If you know that get was called earlier there's no need to call it again.

Reply:
I don't want people to accidentally use this incorrectly, and the overhead isn't that large.

Reply:
People do all kind of clever things and I'd prefer to not see forks of this class (with missing bug fixes, etc) just because of this minor issue.
But I'm not going to insist on that. This can be added later.